src/drivers/screen.c

```diff
@@ -0,0 +1,148 @@
+	if (character == '\n') {
+		int rows = offset / (2 * MAX_COLS);
+		offset = get_screen_offset(MAX_COLS - 1, rows);
+	}
+	// Otherwise write the char and the style to the calculated offset
+	else {
+		vidmem[offset] = character;
+		vidmem[offset + 1] = attribute_byte;
+	}
+
+	// Update the offset to the next character cell which is 2 bytes ahead of our current cell
+	int next_offset = 2;
+	offset += next_offset;
+
+	// Adjust the scrolling when we reach the bottom of the screen
+	offset = handle_scrolling(offset);
+
+	// Update the cursor position
+	set_cursor(offset);
+}
+
+int get_cursor() {
+	// The screen device uses the control register to know what query we want to perform 
+	// and sets the data in the data register. 14 - cursor high byte, 15 - cursor low byte
+	
+	uint8_t cursor_high_byte_query = 14;
+	uint8_t cursor_low_byte_query = 15;
+	
+	port_byte_out(REG_SCREEN_CTRL, cursor_high_byte_query);
+	uint8_t cursor_high_byte = port_byte_in(REG_SCREEN_DATA) ;
+
+	port_byte_out(REG_SCREEN_CTRL, cursor_low_byte_query);
+	uint8_t cursor_low_byte = port_byte_in(REG_SCREEN_DATA);
+
+	int offset = (cursor_high_byte << 8) + cursor_low_byte;
+
+	// The cursor offset we got from the VGA hardware is the number of characters we
+	// need to multiply it by 2 to convert it to a cell offset(each cell has the char and the style)
+	int cell_size = 2;
+	offset *= cell_size;
+
+	return offset;
+}
+
+void set_cursor(const int cell_offset) {
+	// Because each cell in the VGA is of size 2 we need to devide our offset by 2
+	// to convert to char offset from cell offset
+	
+	int cell_size = 2;
+	const int char_offset = cell_offset / cell_size;
+
+	uint8_t cursor_high_byte = (char_offset >> 8) & 0xff;
+	uint8_t cursor_low_byte = char_offset & 0xff;
+
+	uint8_t cursor_high_byte_query = 14;
+	uint8_t cursor_low_byte_query = 15;
+
+	port_byte_out(REG_SCREEN_CTRL, cursor_high_byte_query);
+	port_byte_out(REG_SCREEN_DATA, cursor_high_byte);
+
+	port_byte_out(REG_SCREEN_CTRL, cursor_low_byte_query);
+	port_byte_out(REG_SCREEN_DATA, cursor_low_byte);
+}
+
+void print_at(const char* message, const int col, const int row) {
+	// Update cursor location if col and row aren't negative
+	if (col >= 0 && row >= 0) {
+		set_cursor(get_screen_offset(col, row));
+	}
+
+	// Loop through the string and print it
+	uint32_t i = 0;
+	while (message[i] != 0) {
+		print_char(message[i], col, row, WHITE_ON_BLACK);
+		++i;
+	}
+}
+
+void print(const char* message) {
+	print_at(message, -1, -1);
+}
+
+void clear_screen() {
+	// Loop through the screen and set it's byte to ' '
+	for (uint32_t row = 0; row < MAX_ROWS; row++) {
+		for (uint32_t col = 0; col < MAX_COLS; col++) {
+			print_char(' ', col, row, WHITE_ON_BLACK);
+		}
+	}
+	// Set the cursor back at the top left
+	set_cursor(get_screen_offset(0, 0));
+}
+
+int get_screen_offset(const int col, const int row) {
+	// To calculate we use ((row * MAX_COL) + col) * cell_size
+	int cell_size = 2;
```

Comment:
This is not Hogwarts. Magic isn't allowed here.

Move to define pls

Reply:
wingardium leviosa, fixed